ic10emu/src/vm/object/templates.rs

```diff
@@ -0,0 +1,1409 @@
+                GenericLogicableDeviceConsumerMemoryReadable,
+                GenericLogicableDeviceMemoryReadWriteable, GenericLogicableDeviceMemoryReadable,
+                GenericStorage,
+            },
+            humans::{EntityInfo, HumanPlayer},
+            traits::*,
+            LogicField, Name, Slot, SlotOccupantInfo,
+        },
+        VM,
+    },
+};
+use serde_derive::{Deserialize, Serialize};
+use stationeers_data::{
+    enums::{
+        prefabs::StationpediaPrefab,
+        script::{LogicSlotType, LogicType},
+    },
+    templates::*,
+};
+use strum::{EnumProperty, IntoEnumIterator};
+#[cfg(feature = "tsify")]
+use tsify::Tsify;
+#[cfg(feature = "tsify")]
+use wasm_bindgen::prelude::*;
+
+use super::{stationpedia, MemoryAccess, ObjectID, VMObject};
+
+#[derive(Clone, Debug, PartialEq, PartialOrd, Serialize, Deserialize)]
+#[cfg_attr(feature = "tsify", derive(Tsify))]
+#[cfg_attr(feature = "tsify", tsify(into_wasm_abi, from_wasm_abi))]
+pub enum Prefab {
+    Hash(i32),
+    Name(String),
+}
+
+impl std::fmt::Display for Prefab {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let (known_prefab, unknown_str) = match self {
+            Self::Hash(hash) => (
+                StationpediaPrefab::from_repr(*hash),
+                format!("Unknown({hash}))"),
+            ),
+            Self::Name(name) => (StationpediaPrefab::from_str(&name).ok(), name.clone()),
+        };
+        if let Some(known) = known_prefab {
+            write!(f, "{known}")
+        } else {
+            write!(f, "{unknown_str}")
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(feature = "tsify", derive(Tsify))]
+#[cfg_attr(feature = "tsify", tsify(into_wasm_abi, from_wasm_abi))]
+pub struct ObjectInfo {
+    pub name: Option<String>,
+    pub id: Option<ObjectID>,
+    pub prefab: Option<Prefab>,
+    pub slots: Option<BTreeMap<u32, SlotOccupantInfo>>,
+    pub damage: Option<f32>,
+    pub device_pins: Option<BTreeMap<u32, ObjectID>>,
+    pub connections: Option<BTreeMap<u32, ObjectID>>,
+    pub reagents: Option<BTreeMap<i32, f64>>,
+    pub memory: Option<Vec<f64>>,
+    pub logic_values: Option<BTreeMap<LogicType, f64>>,
+    pub entity: Option<EntityInfo>,
+    pub source_code: Option<String>,
+    pub circuit: Option<ICInfo>,
+}
+
+impl From<&VMObject> for ObjectInfo {
+    fn from(obj: &VMObject) -> Self {
+        let obj_ref = obj.borrow();
+        ObjectInfo {
+            name: Some(obj_ref.get_name().value.clone()),
+            id: Some(*obj_ref.get_id()),
+            prefab: Some(Prefab::Hash(obj_ref.get_prefab().hash)),
+            slots: None,
+            damage: None,
+            device_pins: None,
+            connections: None,
+            reagents: None,
+            memory: None,
+            logic_values: None,
+            entity: None,
+            source_code: None,
+            circuit: None,
+        }
+    }
+}
+
+impl ObjectInfo {
```

Comment:
should probably document these functions

Reply:
👍